quic/s2n-quic-transport/src/stream/controller.rs

```diff
@@ -597,14 +597,22 @@ impl IncomingController {
     }
 
     fn on_remote_open_stream(&mut self, stream_id: StreamId) -> Result<(), transport::Error> {
-        let max_stream_id = StreamId::nth(
+        let allowed_streams = self
+            // the limit of streams to open
+            .max_streams_sync
+            .latest_value()
+            .as_u64()
+            .checked_sub(1)
+            .ok_or(transport::Error::STREAM_LIMIT_ERROR)?;
+        // convert stream index into stream id
+        let max_allowed_stream_id = StreamId::nth(
             stream_id.initiator(),
             stream_id.stream_type(),
-            self.max_streams_sync.latest_value().as_u64(),
+            allowed_streams,
         )
         .expect("max_streams is limited to MAX_STREAMS_MAX_VALUE");
 
-        if stream_id > max_stream_id {
+        if stream_id > max_allowed_stream_id {
```

Comment:
`max_allowed_stream_id` might need to be renamed or a comment added to explain this

Reply:
start with limits: (total number of streams allowed)
`self.max_stream_sync`
calculate max_allowed_stream_id based on limits: (-1 since streams are 0-indexed)
`checked_sub(1)`

compare
stream_id > max_allowed_stream_id

---
if I were to remove `>=` I would need to call the variable `more_than_allowed_stream_id` which seems odd.

stream_id >= more_than_allowed_stream_id

Reply:
I can make the `checked_sub(1)` step separate and add a comment explaining why we get a STREAM_LIMIT_ERROR.. but that does seems confusing.

could we do `!(stream_id < max_allowed_stream_id)` which also seems not ideal.

Reply:
broke the calculation into steps and added comments: https://github.com/aws/s2n-quic/pull/1402/commits/29663503feb4216182f51a7d7c2ba2e0d5c56a9e